src/main/java/org/chameleoncloud/api/admin/UserGroupRolesResource.java

```diff
@@ -0,0 +1,384 @@
+    // and check if a user applies to the policies by comparing the group ids
+
+    SortedMap<Integer, String> result = new TreeMap<Integer, String>();
+    Map<String, Set<String>> policyGroups = new HashMap<String, Set<String>>();
+    permission.getAssociatedPolicies().stream()
+        .filter(policy -> policy.getType().equals(GROUP_POLICY_TYPE)).forEach(policy -> policyGroups
+            .put(policy.getName(), getGroupIdsFromPolicyConfig(policy.getConfig())));
+
+    for (String policy : policyGroups.keySet()) {
+      Set<String> policyGroupIds = policyGroups.get(policy);
+      // get overlap groups between policy groups and user groups
+      policyGroupIds.retainAll(groupIds);
+      if (!policyGroupIds.isEmpty()) {
+        // the user applies to the policy
+        String policyName = UserGroupRoles.parsePolicyName(policy)[0];
+        result.put(UserGroupRoles.POLICY_HIERARCHY.get(policyName), policyName);
+      }
+    }
+
+    return result;
+  }
+
+  private Set<String> getPermissionScopes(Policy permission) {
+
+    // get the scopes of a permission
+
+    return permission.getScopes().stream().map(Scope::getName).collect(Collectors.toSet());
+  }
+
+  private Policy getPolicyByPermissionName(String permissionName, GroupModel group) {
+
+    switch (permissionName) {
+      case "view":
+        return groupPermission.viewPermission(group);
+      case "manage":
+        return groupPermission.managePermission(group);
+      case "view-members":
+        return groupPermission.viewMembersPermission(group);
+      case "manage-members":
+        return groupPermission.manageMembersPermission(group);
+      case "manage-membership":
+        return groupPermission.manageMembershipPermission(group);
+      default:
+        return null;
+    }
+  }
+
+  private Boolean setFineGrainedPermissions(GroupModel group) {
+
+    // check if the group is a root group
+    // if a root group, check if fine grained permission is enabled
+    // enable fine grained admin permission if not
+
+    if (group.getParent() != null) {
+      groupPermission.setPermissionsEnabled(group, false);
+      return false;
+    }
+    if (!groupPermission.isPermissionsEnabled(group)) {
+      groupPermission.setPermissionsEnabled(group, true);
+    }
+    return true;
+  }
+
+  private void grantPermission(String permissionName, GroupModel group, GroupModel subGroup,
+      String policyName) throws RuntimeException {
+
+    // associate the group policy to the scope permissions
+    // if policy does not exist, create one lazily
+
+    Policy groupScopePolicy = getPolicyByPermissionName(permissionName, group);
+    if (groupScopePolicy == null) {
+      throw new RuntimeException(
+          String.format("Can not get policy from fine grained permission %s", permissionName));
+    }
+    String groupId = group.getId();
+    Set<String> associatedPolicies = groupScopePolicy.getAssociatedPolicies().stream()
```

Comment:
One thing that's not clear to me here is the relationship between the group, the policy, and the permission scopes. For example, the manager group has "manage" and "manage-members" scope. Are there two separate policies to grant these two? Or are both scopes part of a single policy? The design spec isn't clear but it seems cleaner to have one policy per "role" we are defining. So the policy would include both scopes in this case. Is that how this is working? I'm not sure how the GroupPermissionManagement thing works.